torch/_decomp/decompositions_for_rng.py

```diff
@@ -0,0 +1,251 @@
+        # the philox_rand functional primitives.
+        if mode == "forward":
+            cls.fwd_state.set_state(seed, offset)
+        else:
+            assert mode == "backward"
+            cls.bwd_state.set_state(seed, offset)
+
+    @classmethod
+    def get_state_as_tensor(cls):
+        # The only reason this exists is because we override get_rng_state and
+        # set_rng_state during tracing. get_rng_state expects a tensor output,
+        # so return (seed, offset) tuple upset other parts of the program like
+        # ctx.saved_tensors.
+
+        # A bad consequence is that if user saves and restores rng state, we
+        # have little bit of ugliness in the generated code, where we first
+        # concat the (seed, offset) to create a tensor for get_rng_state, and
+        # then split it back to get (seed, offset) tuple in set_rng_state.
+
+        # TODO: Investigate if there is be a better way to wrap the tuple in a
+        # false Tensor object, and then desugar it later on.
+        return cls.running_state.get_state_as_tensor()
+
+    @classmethod
+    def get_state_as_tuple(cls):
+        return cls.running_state.get_state_as_tuple()
+
+    @classmethod
+    def set_state_from_tensor(cls, x):
+        # This is only needed because we override set_rng_state. Look at the
+        # comment in get_state_from_tensor method.
+        cls.running_state.set_state_from_tensor(x)
+
+    @classmethod
+    def advance_offset(cls, consumed_offset):
+        cls.running_state.advance_offset(consumed_offset)
+
+    @classmethod
+    def get_current_relative_offset(cls):
+        return cls.running_state.relative_offset
+
+    @classmethod
+    def get_accumulated_offsets(cls):
+        fwd_offset = cls.fwd_state.relative_offset
+        bwd_offset = cls.bwd_state.relative_offset
+        return PhiloxTotalOffsets(fwd_offset, bwd_offset)
+
+
+class CUDARngStateHelper:
+    @staticmethod
+    def get_torch_state_as_tuple(fake_mode=None):
+        if not torch.cuda.is_available():
+            seed = torch.tensor(0)
+            offset = torch.tensor(0)
+            if fake_mode:
+                seed = fake_mode.from_tensor(seed)
+                offset = fake_mode.from_tensor(offset)
+            return seed, offset
+
+        rng_state = torch.cuda.get_rng_state()
+        if fake_mode:
+            rng_state = fake_mode.from_tensor(rng_state)
+        # Rng state is [64-bit seed, 64-bit offset]
+        seed = rng_state[0:8].view(dtype=torch.int64)[0]
+        offset = rng_state[8:].view(dtype=torch.int64)[0]
+        return seed, offset
+
+    @staticmethod
+    def set_torch_state_tensor(seed, offset):
+        # Rng state is [64-bit seed, 64-bit offset]
+        seed_portion = seed.reshape([1]).view(torch.uint8)
+        offset_portion = offset.reshape([1]).view(torch.uint8)
+        new_state = torch.cat([seed_portion, offset_portion])
+        torch.cuda.set_rng_state(new_state)
```

Comment:
You're allocating, like five tensor impls here. This is in the hotpath, and for overhead bound models this could be quite expensive. And there's no need for it; the underlying CUDAGenerator can directly set seed and offset, so you just need to expose an API for it.

Reply:
Relevant PRs - https://github.com/pytorch/pytorch/pull/98965 and https://github.com/pytorch/pytorch/pull/99113